File: 1 - metodosBusca/python/app/core/busca/a_estrela.py
import heapq
from .busca_heuristica import BuscaHeuristica
from .nodo import Nodo

class AEstrela(BuscaHeuristica):
    def __init__(self, mostra_status_console=None):
        super().__init__(mostra_status_console)
        self.max_f = -1  # max F
        self.max_abertos = -1  # max abertos
        self.the_best = None

    def set_max_f(self, m):
        self.max_f = m

    def set_max_abertos(self, m):
        self.max_abertos = m

    def get_the_best(self):
        return self.the_best

    def busca(self, inicial):
        self.status.inicia()
        self.init_fechados()
        
        abertos = []
        heapq.heappush(abertos, (0, Nodo(inicial, None)))
        self.the_best = abertos[0][1]  # o melhor nodo já gerado
        
        while not self.parar and abertos:
            _, melhor = heapq.heappop(abertos)
            self.status.explorando(melhor, len(abertos))
            if melhor.estado.eh_meta():
                self.status.termina(True)
                return melhor
            
            if self.max_f < 0 or melhor.f() < self.max_f:
                for suc in self.sucessores(melhor):
                    heapq.heappush(abertos, (suc.f(), suc))
            if self.max_abertos > 0 and len(abertos) > self.max_abertos:
                break
            
            # o "the best" e o código que segue só para fins de interface
            if melhor.f() < self.the_best.f():
                self.the_best = melhor
        
        self.status.termina(False)
        return None

    def __str__(self):
        return "A* - busca heurística"

File: 1 - metodosBusca/cSharp/AEstrela.java
package busca;

import java.util.PriorityQueue;
import java.util.Queue;

/**
 *   Algoritmos de Busca A*
 *
 *   @author Jomi Fred Hübner
 */
public class AEstrela extends BuscaHeuristica {
    
	int maxF = -1; // max F
    int maxAbertos = -1; // max abertos
	Nodo theBest;

    /** busca sem mostrar status */
    public AEstrela() {
    }
    
    /** busca mostrando status */
    public AEstrela(MostraStatusConsole ms) {
        super(ms);
    }
	
	/** seta o limite para f(), -1 é ilimitado */
	public void setMaxF(int m) {
		maxF = m;
	}

    /** seta o limite para o nro de abertos, -1 é ilimitado */
    public void setMaxAbertos(int m) {
        maxAbertos = m;
    }
	
	public Nodo getTheBest() {
		return theBest;
	}
	
    /**
     *
     * Busca a solução por busca em heurística.
     *                              ----------
     * (baseado no Russel & Norvig)
     */
    public Nodo busca(Estado inicial) {
        status.inicia();
        initFechados();
        
        //Priority_Queue abertos = new Heap(100, getNodoComparatorF()); // lista ordenada por f()
        Queue<Nodo> abertos = new PriorityQueue<Nodo>(100, getNodoComparatorF()); // lista ordenada por f()
        Nodo nInicial = new Nodo(inicial, null);
        abertos.add(nInicial);
        theBest = nInicial; // o melhor nodo já gerado
        
        while (!parar && abertos.size() > 0) {
            
            Nodo melhor = abertos.remove();
            status.explorando(melhor, abertos.size());
            if (melhor.estado.ehMeta()) {
                status.termina(true);
                return melhor;
            }
            
            if (maxF < 0 || melhor.f() < maxF) {
                abertos.addAll( sucessores(melhor) );
            }
            if (maxAbertos > 0 && abertos.size() > maxAbertos) {
                break;
            }
            
            // o "the best" e o código que segue só para fins de interface
            if (melhor.f() < theBest.f()) {
                theBest = melhor;
                //print("\nMelhor (em profundidade "+melhor.getProfundidade()+", h="+((Heuristica)theBest.estado).h()+")="+melhor);
            }
            
        }
        status.termina(false);
        return null;
    }
    
    public String toString() {
    	return "A* - busca heurística"; 
    }
    
}
